source/Octopus.Tentacle.Tests.Integration/WorkspaceCleanerTests.cs

```diff
@@ -0,0 +1,183 @@
+                            {
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should exist before we complete");
+
+                                await Task.Delay(2 * 500, CancellationToken);
+
+                                Directory.Exists(workspaceDirectory).Should().BeTrue("Workspace should not have been cleaned up");
+                            })
+                            .Build())
+                    .Build())
+                .Build(CancellationToken);
+
+            workspaceDirectory = GetWorkspaceDirectoryPath(clientAndTentacle.RunningTentacle.HomeDirectory, startScriptCommand.ScriptTicket.TaskId);
+
+            await clientAndTentacle.TentacleClient.ExecuteScript(startScriptCommand, CancellationToken, null, new InMemoryLog());
+            
+            Directory.Exists(workspaceDirectory).Should().BeFalse("Workspace should be naturally cleaned up after completion");
+        }
+        
+        [Test]
+        [TentacleConfigurations(testDefaultTentacleRuntimeOnly: true)]
+        public async Task WhenTentacleClientCrashesAndTentacleNeverGotToldToComplete_ThenWorkspaceIsConsideredRunning_AndWillNotBeDeleted(TentacleConfigurationTestCase tentacleConfigurationTestCase)
+        {
+            var cleanerDelay = TimeSpan.FromMilliseconds(500);
+            var deleteWorkspacesOlderThan = TimeSpan.FromMilliseconds(500);
+
+            var startScriptCommand = new StartScriptCommandV2Builder().WithScriptBody(b => b.Print("Hello")).Build();
+
+            await using var clientAndTentacle = await tentacleConfigurationTestCase.CreateBuilder()
+                .WithTentacle(b =>
+                {
+                    b.WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.CleanerDelayEnvironmentVariableName, cleanerDelay)
+                        .WithRunTentacleEnvironmentVariable(WorkspaceCleanerConfiguration.DeleteWorkspacesOlderThanTimeSpanEnvironmentVariableName, deleteWorkspacesOlderThan);
+                })
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .DecorateScriptServiceV2With(d =>
+                        d.BeforeCompleteScript((_, _) => throw new NotImplementedException("Force failure to simulate tentacle client crashing, and ensure we do not completion"))
+                        .Build())
+                    .Build())
+                .Build(CancellationToken);
+            
+            await AssertionExtensions
+                .Should(() => clientAndTentacle.TentacleClient.ExecuteScript(startScriptCommand, CancellationToken, null, new InMemoryLog()))
+                .ThrowAsync<NotImplementedException>();
+
+            await Task.Delay(1000, CancellationToken);
+
+            var workspaceDirectory = GetWorkspaceDirectoryPath(clientAndTentacle.RunningTentacle.HomeDirectory, startScriptCommand.ScriptTicket.TaskId);
+
+            Directory.Exists(workspaceDirectory).Should().BeTrue();
```

Comment:
I'm confused!! Shouldn't this be deleted?

Reply:
Ahh ok I see, the workspace remains in memory until Tentacle is restarted so we don't clean up!!

I think you have caught a missed requirement here. I think we would want to clean this up! Tentacle could run for weeks or months which would render our cleanup job redundant!!

Reply:
I fixed this up. Now, if it finds a running script, it also queries the `State`.

You mentioned maybe asking "when it stopped". But I decided against that because if Tentacle dies, then the in-memory date would die. So we would still need a way of knowing how old something is (i.e. looking at the log date), so we may as well use the same mechanism for both. Also, this way means not adding extra code to the existing script service 😁